### PR TITLE
Provide assigns to arity 2 fun in update/3

### PR DESCRIPTION
**Pull request by 03juan**
Curries fun/2 with assigns for both socket_or_assigns arguments.

Complement to PR#1997

### CODE REVIEW

File: lib/phoenix_live_view.ex [hunk 1]
```diff
@@ -875,28 +875,41 @@ defmodule Phoenix.LiveView do
   The update function receives the current key's value and
   returns the updated value. Raises if the key does not exist.
 
+  The update function may also be of arity 2, in which case it receives
+  the current key's value as the first argument and the current assigns
```
Comment (03juan) [inline lib/phoenix_live_view.ex:879]: I reused the grammar from the previous paragraph in "the current key's value", but they may read better as "the key's current value". Any interesting in updating them at this opportunity?

Comment (03juan) [reply lib/phoenix_live_view.ex:879]: > [selected|provided

I feel this is implied by the function's context, but happy to follow any recommendation from the maintainers.

File: test/phoenix_live_view_test.exs [hunk 2]
```diff
@@ -342,6 +342,22 @@ defmodule Phoenix.LiveViewUnitTest do
     end
   end
 
+  describe "update with arity 2 function" do
+    test "passes socket assigns to update function" do
+      socket = @socket |> assign(key: "value", key2: "another") |> Utils.clear_changed()
+
+      socket = update(socket, :key2, fn _, %{key: key} -> key end)
```
Comment (leandrocp) [inline test/phoenix_live_view_test.exs:349]: It might be worth adding a test using both args to cover all cases.

**Review by leandrocp — COMMENTED** [review]
Just two minor suggestions, not blocking. LGTM

File: test/phoenix_live_view_test.exs (hunk 2 above)
Comment (03juan) [reply test/phoenix_live_view_test.exs:349]: Makes sense. Thank you

File: lib/phoenix_live_view.ex [hunk 3]
```diff
@@ -875,28 +875,41 @@ defmodule Phoenix.LiveView do
...
+  as the second argument. Raises if the key does not exist.
+
   ## Examples
 
       iex> update(socket, :count, fn count -> count + 1 end)
       iex> update(socket, :count, &(&1 + 1))
+      iex> update(socket, :double_count, fn _, %{count: count} -> count * 2 end)
```
[the remark this answers is not in the document]
Comment (03juan) [reply lib/phoenix_live_view.ex:886]: I like replacing the `:double_count` line with the `:users_count` example, or do you think we should keep both?

[the remark this answers is not in the document]
Comment (leandrocp) [reply lib/phoenix_live_view.ex:886]: I'm okay with just the `:users_count` example, it does cover both cases already.

[the remark this answers is not in the document]
Comment (03juan) [reply lib/phoenix_live_view.ex]: Your suggestion wasn't using the current value so I went with something that did, but thank you for the tip.

File: lib/phoenix_live_view.ex [hunk 4]
```diff
@@ -875,28 +875,41 @@ defmodule Phoenix.LiveView do
...
   ## Examples
 
       iex> update(socket, :count, fn count -> count + 1 end)
       iex> update(socket, :count, &(&1 + 1))
+      iex> update(socket, :max_users_this_session, fn current_max, %{users: users} -> max(current_max, length(users)) end)
   """
   def update(socket_or_assigns, key, fun)
 
+  def update(%Socket{assigns: assigns} = socket, key, fun) when is_function(fun, 2) do
+    update(socket, key, &fun.(&1, assigns))
```
Comment (leandrocp) [inline lib/phoenix_live_view.ex:891]: Wait a minute :)
Here `assigns` is passing reserved keys like `:__changed__`, `:flash` and others which IMO should be stripped out as those keys are private.

You could call https://hexdocs.pm/phoenix_live_view/Phoenix.LiveView.Helpers.html#assigns_to_attributes/2 to clean up `assigns`

Comment (josevalim) [reply lib/phoenix_live_view.ex:891]: No, it is fine to preserve those keys. `assigns` will always have those keys anyway, anywhere, anytime. :)

Comment (leandrocp) [reply lib/phoenix_live_view.ex:891]: All right, even simpler 👍🏻

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**